src/polymer/root-selector-to-html.ts

```diff
@@ -0,0 +1,106 @@
+/**
+ * @license
+ * Copyright (c) 2018 The Polymer Project Authors. All rights reserved.
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ * The complete set of authors may be found at
+ * http://polymer.github.io/AUTHORS.txt
+ * The complete set of contributors may be found at
+ * http://polymer.github.io/CONTRIBUTORS.txt
+ * Code distributed by Google as part of the polymer project is also
+ * subject to an additional IP rights grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+import * as dom5 from 'dom5';
+import {Document, ParsedCssDocument, Severity, Warning} from 'polymer-analyzer';
+import * as shady from 'shady-css-parser';
+
+import {registry} from '../registry';
+import {Rule} from '../rule';
+import {getDocumentContaining, stripIndentation, stripWhitespace} from '../util';
+
+const p = dom5.predicates;
+const isCustomStyle = p.AND(
+    p.hasTagName('style'),
+    p.OR(
+        p.hasAttrValue('is', 'custom-style'),
+        (node: dom5.Node) => !!(
+            node.parentNode && p.hasTagName('custom-style')(node.parentNode))));
+const isElementStyle = p.AND(
+    p.hasTagName('style'),
+    (node: dom5.Node) => !!(
+        node.parentNode && node.parentNode.nodeName === '#document-fragment'));
+
+class RootSelectorToHtml extends Rule {
+  code = 'root-selector-to-html';
+  description = stripIndentation(`
+      Warns when using :root inside an element's template or custom-style
+  `);
+
+  async check(document: Document) {
+    return [
+      ...this.generateWarnings(
+          document, isCustomStyle, 'html'),  // Check custom styles
+      ...this.generateWarnings(
+          document, isElementStyle, ':host > *'),  // Check element styles
```

Comment:
If it's a dom-module without an element then it might be a style which is intended for inclusion with include, in which case we want to convert it to _both_ html and :host. See, e.g. https://github.com/PolymerElements/paper-styles/blob/master/element-styles/paper-material-styles.html#L37

If there's other parts of the selector (does that make sense in this context? I'm not sure), we also need to preserve them, so that `:root foo` becomes `:host > * foo, html foo`

Reply:
hmmmm good call on the `dom-module` part, I will see if I can get that fixed as well.

Reply:
In that case, let's not fix except in the very simple case of just `:root`